latin_squares.py

```diff
@@ -133,226 +129,196 @@ def resample(self):  # TODO
+            self.calc_weight()  # Recalculate weights
             print(f"Iteration {self.iteration} done!")
-            print(f"λ_{self.iteration} : {self.lambdas[-1]}")
-
+            print(f"λ_{self.iteration} : {self.lambd}")
             self.calc_log_normalizing_constant_and_update()
+        print('SMC finished!')
 
-            self.iteration += 1
-
-    def calc_log_normalizing_constant_and_update(self):
+    def calc_log_normalizing_constant_and_update(self):  # DONE
         """
         See pg 305 of Papaspiliopoulos / Chopin.
         I cross referenced with the `particles` library by Chopin.
 
-        We can caluculate logLt by 
+        We can caluculate logLt by
         logLt = \sum_{s=0}^{t} log( \sum_{n=1}^{N} w_s^n )
 
         So for every iteration, we add calculate the log normalising constant
         and add it to `self.LogLt`.
         """
-        self.logLt += np.log( np.mean(np.exp(self.unnormalized_logweights)) )
-
+        self.logLt += np.log(np.mean(np.exp(self.w_log)))
 
 
-def sample(d):
+def sample(d, seed=None):  # DONE
     """
     Sample a permutation of 0, 1, ..., d-1
     """
-    return np.random.choice(range(d), size=d, replace=False)
+    if seed is not None:
+        np.random.seed(seed)
+    return np.random.permutation(d)
 
 
-# TODO I'm not sure if this is actually correct
-def sample_matrix(d):
-    """
-    Sample a d x d matrix where every row is a permutation of
-    0, 1, ..., d-1
-    """
-    return np.matrix([sample(d) for _ in range(d)])
+class LatinKernel:  # DONE
+    def __init__(self):
+        pass
 
-
-def latin_kernel(x):
-    """
-    Takes a d x d matrix and selects a row i
-    and two columns j1 and j2 at random.
-    then it swaps the values of x[i,j1] and x[i,j2]
-    """
-    d = x.shape[0]
-    m = x.copy()
-    i, j1, j2 = np.random.randint(low=0, high=d, size=3)
-    x1 = x[i, j1]
-    x2 = x[i, j2]
-    m[i, j1] = x2
-    m[i, j2] = x1
-    return m
+    @staticmethod
+    def sample(x_cur):
+        """
+        Takes a d x d matrix and selects a row i
+        and two columns j1 and j2 at random.
+        then it swaps the values of x[i,j1] and x[i,j2]
+        """
+        assert x_cur.shape[0] == x_cur.shape[1]
+        d = x_cur.shape[0]
+        x_new = deepcopy(x_cur)
+        i = np.random.choice(d)
+        j1, j2 = np.random.choice(d, size=2, replace=False)
+        x_new[i, j1], x_new[i, j2] = x_new[i, j2], x_new[i, j1]
+        return x_new
 
 
-def V_latin(x):
+def V_latin(x):  # DONE
     """
     Calculate score of latin square.
     """
     d = x.shape[1]
-    return (
-        sum(
-            (
-                sum((x[i, j] == 1) for i in range(d)) ** 2
-                for l in range(d)
-                for j in range(d)
-            )
-        )
-        - d ** 2
-    )
+    return sum(sum(sum(x[i, j] == l for i in range(d))**2 for l in range(d)) - d for j in range(d))
```

Comment:
perhaps
`
sum(sum(sum(x[i, j] == l for i in range(d))**2 for l in range(d)) for j in range(d)) - d**2 is better? or maybe it doesn't make a difference idk
`

Reply:
yeah, might be, idk whether it gives any computational gain